### ISSUE TITLE
rounding a decimal

### ISSUE THREAD

**Issue by mehdigmira** [python/mypy#6211]
With the latest version of mypy this causes an error:

```python
from decimal import Decimal

x = Decimal('10.11111')

y = round(x, 2)

print(y)

```

```
commands/debug.py:5: error: No overload variant of "round" matches argument types "Decimal", "int"
commands/debug.py:5: note: Possible overload variants:
commands/debug.py:5: note:     def round(number: float, ndigits: None) -> int
commands/debug.py:5: note:     def round(number: float, ndigits: int) -> float
commands/debug.py:5: note:     <2 more similar overloads not shown, out of 6 total overloads>

```

while running the script does not raise an error.

**Comment by gvanrossum** [comment]
This is a regression since 0.650. I can repro this in 0.660, and it doesn't look like a typeshed bug. In typeshed the builtin `round()` function is defined as taking any argument that follows the `SupportsRound` protocol, and `Decimal` does indeed define `__round__`. So what prevents mypy from understanding that? Maybe that it's defined as an overload? @Michael0x2a

**Comment by Michael0x2a** [comment]
I did some bisecting, and it looks like this is a regression introduced by https://github.com/python/typeshed/pull/2674. 

That commit modified the the signature of `Decimal.__round__` from:

    class Decimal:
        # ...snip...
        def __round__(self, ndigits: Optional[int] = None) -> int: ...

...into:

    class Decimal:
        # ...snip...
        @overload
        def __round__(self) -> int: ...
        @overload
        def __round__(self, ndigits: None) -> int: ...
        @overload
        def __round__(self, ndigits: int) -> Decimal: ...

While this signature is more accurate, it regrettably makes Decimal no longer compatible with the definition of SupportsRound, which looks like this:

    @runtime
    class SupportsRound(Protocol[_T_co]):
        @abstractmethod
        def __round__(self, ndigits: int = ...) -> _T_co: ...

This is because the output of `Decimal.__round__()` differs from the output of `SupportsRound[Decimal].__round__()` -- the former returns int; the latter will return `Decimal`.

This prevents decimals from being a valid match for any of the `round(...)` overloads causing the error above.

---

In this particular case, I think the definition of `SupportsRound` is actually incorrect here and ought to be changed to also use an overload. This same issue [actually came up before](https://github.com/python/typeshed/issues/2638), and I made the same suggestion before (though I didn't remember I did until I started digging into this).

I think the other takeaway here is that we currently don't have a good way of making sure some change to a class in typeshed won't accidentally make it incompatible with some protocol. Maybe we should start making classes like Decimal explicitly inherit the relevant protocols in typeshed whenever possible? Or maybe we try and figure out a way of adding custom unit tests to typeshed?

**Comment by gvanrossum** [comment]
I like the idea of explicitly inheriting better than unit tests -- the
explicit inheritance should complain when the protocol isn't satisified
without having to write any extra code.

**Comment by msullivan** [comment]
Fixed by https://github.com/python/typeshed/pull/2907

**Comment by dms-cat** [comment]
Was this included in v0.701? If so, how do I use it? I've got the following code:

    def calculate(self, raw_value: Decimal, decimal_places: Decimal) -> Dict[str, Decimal]:
        return {self.output_variable_name: round(raw_value, ndigits=int(decimal_places))}

and I'm still getting

> error: No overload variant of "round" matches argument types "Decimal", "int"

**Comment by ilevkivskyi** [comment]
> Was this included in v0.701?

This was _not_ included. Wait for the next release (in a week or two).

**Comment by lezwon** [comment]
Hey Folks, I have a similar error with the following code: https://mypy-play.net/?mypy=latest&python=3.10&gist=b0b3314bb23f3a406ae2d6ced9027721

```python
from numbers import Number
a: Number = 6
int(a)
```

mypy says:
```
main.py:2: error: Incompatible types in assignment (expression has type "int", variable has type "Number")
main.py:3: error: No overload variant of "int" matches argument type "Number"
main.py:3: note: Possible overload variants:
main.py:3: note:     def __new__(cls, Union[str, bytes, SupportsInt, SupportsIndex, SupportsTrunc] = ...) -> int
main.py:3: note:     def __new__(cls, Union[str, bytes], base: SupportsIndex) -> int
```

Shouldn't 'int()` support `Number` args?

**Comment by JelleZijlstra** [comment]
@lezwon please avoid posting on closed issues. You're running into #3186.
